torch/csrc/Exceptions.h

```diff
@@ -138,4 +157,27 @@ struct ValueError : public PyTorchError {
   }
 };
 
+// ATen warning handler for Python
+struct PyWarningHandler{
+public:
+  using warning_buffer_t =
+    std::queue<std::pair<c10::SourceLocation, std::string>>;
+
+  static void py_warning_handler(
+    const c10::SourceLocation& source_location,
+    const std::string& msg);
+
+  static warning_buffer_t warning_buffer;
```

Comment:
This struct is all static members. Any reason for it to be a struct? A namespace would have worked equally as well. You can move the `warning_buffer` declaration to be private to `torch/csrc/Exceptions.cpp` as it doesn't seem to be used anywhere else.

Reply:
Is `warning_buffer` thread safe? I don't think `std::queue` is thread safe. So what lock protects it?